### ISSUE TITLE
Optimization of bssdf parameters leads to memory crash

### ISSUE THREAD

**Issue by osylum** [mitsuba-renderer/mitsuba3#643]
## Summary

I have an optimization loop to optimize bssdf parameters. During the loop, at the first iteration, the program crashes with memory exhaustion error both with cuda and llvm, the later just taking much more time before finishing with the error.

## System configuration

  OS: Windows-10
  CPU: Intel64 Family 6 Model 165 Stepping 5, GenuineIntel
  GPU: NVIDIA RTX A4000
  Python: 3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit (AMD64)]
  NVidia driver: 517.40
  CUDA: 10.0.130
  LLVM: 15.-1.-1

  Dr.Jit: 0.4.1
  Mitsuba: 3.2.1
     Is custom build? False
     Compiled with: MSVC 19.34.31942.0
     Variants:
        scalar_rgb
        scalar_spectral
        cuda_ad_rgb
        llvm_ad_rgb

## Description

I am trying to optimize bssdf parameters (sigma_t, albedo, g). I have only a few renders with low resolution (128) and low sample count (256). The optimization loop is similar as in: https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/radiance_field_reconstruction.html . I tried with cuda and with llvm. 

It failed at the first iteration after throwing this message when using cuda or llvm (takes just longer but produces the same result):

jit_flush_malloc_cache(): Dr.Jit exhausted the available memory and had to flush its allocation cache to free up additional memory. This is an expensive operation and will have a negative effect on performance. You may want to change your computation so that it uses less memory. This warning will only be displayed once.

I am only optimizing 3 parameter and have low resolution images. I don't understand why it exhausts the memory or get stuck. Can you spot something I am doing wrong?

Note that the creation of the scene and rendering from it works fine. The image at first iteration is properly displayed. It fails right after print(f"loss: {loss[0]}") before dr.backward(loss). I tried also a similar example to optimize brdf parameters and it worked fine, both with llvm and cuda; the latter producing a significant speed up.

Thank you

## Steps to reproduce

```python

def optimize(dataset_path):
	"""
	Optimization of bssdf parameters
	"""

	render_resolution = [128,128]
	num_stages = 50

	output_path = os.path.join(dataset_path, 'outputs')
	if os.path.exists(output_path):
		os.makedirs(output_path)

	# get train and test infos
	# list_meshmodels = ['armadillo', 'buddha', 'bun', 'bunny', 'bust', 'cap', 'cube', 'dragon', 'lucy', 'soap', 'star_smooth']
	list_sigma_t = [30., 36., 44., 54., 67., 82., 100., 122., 150., 184., 225., 276.]
	list_albedo = [0.39, 0.59, 0.74, 0.87, 0.95]
	list_g = [0., 0.2, 0.4, 0.6, 0.8]
	meshmodel = 'armadillo'
	sigma_t_target = 44.
	albedo_target = 0.87
	g_target = 0.2
	print(f"{sigma_t_target=}")
	print(f"{albedo_target=}")
	print(f"{g_target=}")
	dataset_infos = get_optimizeset_infos(meshmodel, sigma_t_target, albedo_target, g_target)

	sigma_t_init = (list_sigma_t[0] + list_sigma_t[-1]) * 0.5
	albedo_init = (list_albedo[0] + list_albedo[-1]) * 0.5
	g_init = (list_g[0] + list_g[-1]) * 0.5
	print(f"{sigma_t_init=}")
	print(f"{albedo_init=}")
	print(f"{g_init=}")


	# create scene and emitter transformations
	# need prbvolpath for backpropagating gradients
	nsamples = 256 # 4096
	emitters_to_world = {}
	for i, sample_infos in enumerate(dataset_infos):
		angle = sample_infos["angle"]
		emitters_to_world[str(angle)] = create_emitter_to_world(angle)
	scene, _ = create_scene(meshmodel, -150.,
						   sigmaT = sigma_t_target, albedo = albedo_target, g = g_target,
						   nsamples = nsamples, render_resolution = render_resolution,
						   integrator_type = 'prbvol')

	# access param and define keys to change params values
	scene_params = mi.traverse(scene)
	print(scene_params)
	key_sigma_t = 'object.interior_medium.sigma_t.value.value'
	key_albedo = 'object.interior_medium.albedo.value.value'
	key_g = 'object.interior_medium.phase_function.g'


	# generate reference images
	list_angle = np.array([0., -30., -60., -90., -120., -150., -180.])
        images_true = []
        for i, sample_infos in enumerate(dataset_infos):
	        angle = sample_infos["angle"]
	        scene_params["emitter.to_world"] = emitters_to_world[str(angle)]
	        image = mi.render(scene, scene_params) # TensorXf
	        images_true.append(image)
	        if True:
		        plt.axis('off')
		        plt.imshow(dr.clip(mi.TensorXf(image), 0.0, 1.0))
		        plt.title('reference')
		        plt.show()


	# set initial bssdf parameters
	scene_params[key_sigma_t] = sigma_t_init
	scene_params[key_albedo] = albedo_init
	scene_params[key_g] = g_init
	scene_params.update()

	# render initial scene
	image = mi.render(scene, scene_params)
	plt.axis('off')
	plt.imshow(dr.clip(mi.TensorXf(image), 0.0, 1.0))
	plt.title('initial')
	plt.show()

	# create optimizer
	lr_rate = 0.0001
	optimizer = mi.ad.Adam(lr=lr_rate)
	optimizer[key_sigma_t] = scene_params[key_sigma_t] # needs to be differentiable
	optimizer[key_albedo] = scene_params[key_albedo] # needs to be differentiable
	optimizer[key_g] = scene_params[key_g] # needs to be differentiable
	scene_params.update(optimizer)

	num_stages = 2
	num_iterations_per_stage = 1

	losses = []
	intermediate_images = []
	start_time = timer()
	for stage in range(num_stages):
		print(f"Stage {stage + 1:02d}")

		start_time_stage = timer()
		for it in range(num_iterations_per_stage):
			print(f"Iter {it}")
			total_loss = 0.0
			images = []

			for i, sample_infos in enumerate(dataset_infos):
				print(f"Image {i}")
				# render scene with current params and light direction
				print("render scene")
				angle = sample_infos["angle"]
				scene_params["emitter.to_world"] = emitters_to_world[str(angle)]
				image= mi.render(scene, scene_params)  # passing params is important here

				if True:
					plt.axis('off')
					plt.imshow(dr.clip(mi.TensorXf(image), 0.0, 1.0))
					plt.title('iter: {it}, image {i}')
					plt.show()

				# compute loss
				print("compute loss")
				loss = dr.mean(dr.abs(image - images_true[i]))
				print(f"loss: {loss[0]}")
				dr.backward(loss)
				total_loss += loss[0]

				# Store images at the end of every stage
				if it == num_iterations_per_stage - 1:
					dr.eval(image)
					images.append(image)

			losses.append(total_loss)
			optimizer.step()
			scene_params.update(optimizer)

		print(f"{scene_params[key_sigma_t]=}")
		print(f"{scene_params[key_albdo]=}")
		print(f"{scene_params[key_g]=}")

		intermediate_images.append(images)

		elapsed_time_stage = timer() - start_time_stage
		print('elapsed time for this stage:', elapsed_time_stage)

	elapsed_time_total = timer() - start_time
	print('total elapsed time:', elapsed_time_total)

	print(f"final sigma_t: {scene_params[key_sigma_t]}")
	print(f"final albedo: {scene_params[key_albdo]}")
	print(f"final g: {scene_params[key_g]}")

	# show final result
	final_images = []
	for i, sample_infos in enumerate(dataset_infos):
		angle = sample_infos["angle"]
		scene_params["emitter.to_world"] = emitters_to_world[str(angle)]
		image = mi.render(scene, scene_params)
		final_images.append(image)
	for stage, inter in enumerate(intermediate_images):
		plot_list(inter, f'Stage {stage}')
	plot_list(final_images, 'Final')
	plot_list(images_true, 'Reference')

	fig, axs = plt.subplots(1, 2, figsize=(10, 4))
	axs[0].imshow(mi.util.convert_to_bitmap(final_images[1]))
	axs[0].set_title('Reconstructed')
	axs[0].axis('off')
	axs[1].imshow(mi.util.convert_to_bitmap(images_true[1]))
	axs[1].set_title('Reference')
	axs[1].axis('off')
	plt.show()
```

**Comment by njroussel** [comment]
Hi @osylum 

Nothing jumps out as strange to me.

If you can compute the loss, then its the adjoint pass that is consuming too much memory (call to `dr.backward`). I would double check your integrator - you should be using `prbvolpath`.  The memory consumption should scale linearly with your number of rays during the adjoint rendering, so maybe verify your spp and resolution too.

**Comment by Disgruntoad** [comment]
@osylum , I'm running into a similar error. Did you ever resolve this issue?
